src/workerd/api/encoding.c++

```diff
@@ -460,29 +461,203 @@ kj::Maybe<jsg::JsString> TextDecoder::decodePtr(
+  out[3] = static_cast<char>(0x80 | (codepoint & 0x3F));
+}
+
+// Convert UTF-16 with potentially invalid surrogates to UTF-8.
+// Invalid surrogates are replaced with U+FFFD.
+void convertInvalidUtf16ToUtf8(kj::ArrayPtr<const char16_t> input, kj::ArrayPtr<char> out) {
+  size_t position = 0;
+  bool pendingSurrogate = false;
+
+  for (size_t i = 0; i < input.size(); i++) {
+    char16_t c = input[i];
+
+    if (pendingSurrogate) {
+      if (isTrailSurrogate(c)) {
+        encodeSurrogatePair(input[i - 1], c, out.slice(position, out.size()));
+        position += 4;
+        pendingSurrogate = false;
+      } else {
+        position += encodeUtf8CodeUnit(0xFFFD, out.slice(position, out.size()));
+        if (!isLeadSurrogate(c)) {
+          position += encodeUtf8CodeUnit(c, out.slice(position, out.size()));
+          pendingSurrogate = false;
+        }
+      }
+    } else if (isLeadSurrogate(c)) {
+      pendingSurrogate = true;
+    } else {
+      if (isTrailSurrogate(c)) {
+        position += encodeUtf8CodeUnit(0xFFFD, out.slice(position, out.size()));
+      } else {
+        position += encodeUtf8CodeUnit(c, out.slice(position, out.size()));
+      }
+    }
+  }
+
+  if (pendingSurrogate) {
+    encodeUtf8CodeUnit(0xFFFD, out.slice(position, out.size()));
+  }
+}
+
 }  // namespace
 
-jsg::BufferSource TextEncoder::encode(jsg::Lock& js, jsg::Optional<jsg::JsString> input) {
-  auto str = input.orDefault(js.str());
-  auto view = JSG_REQUIRE_NONNULL(jsg::BufferSource::tryAlloc(js, str.utf8Length(js)), RangeError,
-      "Cannot allocate space for TextEncoder.encode");
-  [[maybe_unused]] auto result = encodeIntoImpl(js, str, view);
-  KJ_DASSERT(result.written == view.size());
-  return kj::mv(view);
+jsg::Ref<TextEncoder> TextEncoder::constructor(jsg::Lock& js) {
+  return js.alloc<TextEncoder>();
+}
+
+jsg::JsUint8Array TextEncoder::encode(jsg::Lock& js, jsg::Optional<jsg::JsString> input) {
+  jsg::JsString str = input.orDefault(js.str());
+  std::shared_ptr<v8::BackingStore> backingStore;
+  size_t utf8_length = 0;
+
+  // Fast path: check if string is one-byte before creating ValueView
+  if (str.isOneByte(js)) {
+    auto length = str.length(js);
+    // Allocate buffer for Latin-1. Use v8::ArrayBuffer::NewBackingStore to avoid creating
+    // JS objects during conversion.
+    backingStore = js.allocBackingStore(length, jsg::Lock::AllocOption::UNINITIALIZED);
+    auto backingData = reinterpret_cast<kj::byte*>(backingStore->Data());
+
+    str.writeOneByte(js, kj::arrayPtr(backingData, length));
+
+    utf8_length =
+        simdutf::utf8_length_from_latin1(reinterpret_cast<const char*>(backingData), length);
+
+    if (utf8_length == length) {
+      // ASCII fast path: no conversion needed, Latin-1 is same as UTF-8 for ASCII
+      auto array = v8::Uint8Array::New(v8::ArrayBuffer::New(js.v8Isolate, backingStore), 0, length);
+      return jsg::JsUint8Array(array);
+    }
```

Comment:
to be safe, a `KJ_DASSERT` that verifies that `utf8_length > length` would be good here.